src/shim.c

```diff
@@ -151,97 +155,287 @@ err_exit(const char *format, ...)
+ *
+ * \param shim \ref cc_shim
+ * \param fr \ref frame
  *
- * \param fd File descriptor to send the message to(should be proxy ctl socket fd)
- * \param Hyperstart cmd id
- * \param json Json payload
+ * \return true on success, false otherwise
  */
-void
-send_proxy_hyper_message(int fd, const char *hyper_cmd, const char *json) {
-	char      *proxy_payload = NULL;
-	char      *proxy_command_id = "hyper";
-	char      *proxy_ctl_msg = NULL;
-	size_t     len = 0, offset = 0;
+bool
+write_frame(struct cc_shim *shim, struct frame *fr)
+{
+	size_t     offset = 0;
 	ssize_t    ret;
+	ssize_t    total_size = 0;
 
-	/* cc-proxy has the following format for "hyper" payload:
-	 * {
-	 *    "id": "hyper",
-	 *    "data": {
-	 *        "hyperName": "ping",
-	 *        "data": "json payload",
-	 *    }
-	 * }
-	*/
-
-	if ( !json || fd < 0) {
-		return;
+	if (! (shim && fr)) {
+		return false;
 	}
 
+	uint8_t *msg = serialize_frame(fr, &total_size);
 
-	ret = asprintf(&proxy_payload,
-			"{\"id\":\"%s\",\"data\":{\"hyperName\":\"%s\",\"data\":%s}}",
-			proxy_command_id, hyper_cmd, json);
-
-	if (ret == -1) {
-		abort();
+	if (! msg ) {
+		return false;
 	}
 
-	proxy_ctl_msg = get_proxy_ctl_msg(proxy_payload, &len);
-	free(proxy_payload);
-
-	while (offset < len) {
-		ret = write(fd, proxy_ctl_msg + offset, len-offset);
+	while (offset < total_size) {
+		ret = write(shim->proxy_sock_fd, msg + offset, (size_t)total_size-offset);
 		if (ret == EINTR) {
 			continue;
 		}
 		if (ret <= 0 ) {
-			free(proxy_ctl_msg);
+			free(msg);
 			shim_error("Error writing to proxy: %s\n", strerror(errno));
-			return;
+			return false;
 		}
 		offset += (size_t)ret;
 	}
-	free(proxy_ctl_msg);
+	free(msg);
+	return true;
+}
+
+/*!
+ * Send message to proxy.
+ *
+ * \param shim \ref cc_shim
+ * \param type Type of message
+ * \param opcode Opcode of message type
+ * \param payload Payload to be sent
+ * 
+ * \return true on success, false otherwise
+ */
+bool
+send_proxy_message(struct cc_shim *shim, uint8_t type, uint8_t opcode,
+			const char *payload)
+{
+	struct frame fr = { 0 };
+	bool ret;
+
+	if ( !shim || shim->proxy_sock_fd < 0) {
+		return false;
+	}
+
+	fr.header.version = PROXY_VERSION;
+	fr.header.header_len = MIN_HEADER_WORD_SIZE;
+	fr.header.type = type;
+	fr.header.opcode = opcode;
+
+	shim_debug("Sending frame of type:%d, opcode:%d, payload:%s\n",
+			fr.header.type,
+			fr.header.opcode,
+			payload? payload:"(empty)");
+
+	if (payload) {
+		fr.header.payload_len = (uint32_t)strlen(payload);
+		fr.payload = (uint8_t*)payload;
+	}
+
+	ret = write_frame(shim, &fr);
+	return ret;
+}
+
+/*!
+ * Send Connect command to proxy.
+ *
+ * \param shim \ref cc_shim
+ *
+ * \return true on success, false otherwise
+ */
+bool
+send_connect_command(struct cc_shim *shim)
+{
+	char *payload = NULL;
+	bool ret;
+
+	if (! shim) {
+		return false;
+	}
+
+	if ((! shim->token) || (shim->proxy_sock_fd < 0)) {
+		return false;
+	}
+
+	strip_double_quotes(shim->token);
```

Comment:
I think it would be better to just fail if `shim->token` contains any non-alphanumerics. That admittedly does mean the shim would then have knowledge of the token format (rather than just passing the data through), but it would be a quick and safe approach that only lets through data in the expected format.